ODIN_II/SRC/netlist_create_from_ast.cpp

```diff
@@ -3469,51 +3473,129 @@ signal_list_t* create_operation_node(ast_node_t* op, signal_list_t** input_lists
                 add_input_port_information(operation_node, input_port_width);
                 /* hookup the input pins */
-                hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_port_width, verilog_netlist);
+                /* 
+                 * [NOTE] The below function does not handle the signed variables. 
+                 * Indeed. instead of adding 0 as the extension bits, it should be the sign bit wether 1 or 0 
+                 */
+                hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[k], 0, input_port_width, verilog_netlist);
+                current_idx += input_port_width;
             }
-        } else if (input_port_width != -2) {
-            /* IF taking port widths based on preset */
-            /* allocate the pins needed */
-            allocate_more_input_pins(operation_node, input_port_width);
-            /* record this port size */
-            add_input_port_information(operation_node, input_port_width);
 
-            /* hookup the input pins = will do padding of zeros for smaller port */
-            hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_port_width, verilog_netlist);
-            current_idx += input_port_width;
         } else {
-            /* ELSE if taking the port widths as they are */
-            /* allocate the pins needed */
-            allocate_more_input_pins(operation_node, input_lists[i]->count);
-            /* record this port size */
-            add_input_port_information(operation_node, input_lists[i]->count);
+            /* record the size of the shift */
+            int shift_size = op->children[1]->types.vnumber->get_value();
+
+            switch (operation_node->type) {
+                case SL:
+                case ASL: {
+                    /* connect ZERO to outputs that don't have inputs connected */
+                    for (i = 0; i < shift_size; i++) {
+                        // connect 0 to lower outputs
+                        npin_t* zero_pin = allocate_npin();
+                        add_fanout_pin_to_net(verilog_netlist->zero_net, zero_pin);
+
+                        add_pin_to_signal_list(return_list, zero_pin);
+                        zero_pin->node = NULL;
+                    }
 
-            /* hookup the input pins */
-            hookup_input_pins_from_signal_list(operation_node, current_idx, input_lists[i], 0, input_lists[i]->count, verilog_netlist);
+                    /* connect inputs to outputs */
+                    for (i = 0; i < output_port_width - shift_size; i++) {
+                        // connect higher output pin to lower input pin
+                        add_pin_to_signal_list(return_list, input_lists[0]->pins[i]);
+                        input_lists[0]->pins[i]->node = NULL;
+                    }
+                    break;
+                }
+                case SR: {
```

Comment:
you did not address my comment! merge the SR and the ASR logic together.
the only time you replicate the pad is when the value is signed AND it is ASR.
if the value is not signed then it becomes an SR logic. 
either you do the detection of this earlier and if the LHS is not signed, convert ASR to SR during AST_OPTIMIZATION.
or you detect this here and merge the logic

Reply:
I recommend you do the detection earlier and also resize the input based on the output width there, it will make you life simpler since then you can just create the pins here and you can get the assignment size when you resolve assignment sizes